docs/design/2021-02-01-foreign-data-wrapper.md

```diff
@@ -0,0 +1,164 @@
+<!--
+This is a template for TiDB's change proposal process, documented [here](./README.md).
+-->
+
+# Proposal: Foreign Data Wrappers
+
+
+- Author(s):  [iguory](https://github.com/iguoyr)，[wph95](https://github.com/wph95)，[lfkdsk](https://github.com/lfkdsk)
+- Last updated: 2020-02-01
+- Discussion at: https://github.com/pingcap/tidb/issues/18677
+
+## Abstract
+
+FDW（FOREIGN DATA WRAPPER）是一种让数据库可以访问远程数据存储的外部数据包装器。
+
+本提案将遵循 [SQL/MED](https://wiki.postgresql.org/wiki/SQL/MED#Version_4)，通过对 Table Lifecycle 注入 Hook（复用 TiDB plan，在执行器 builder 时候使用自定义算子逻辑），使用 TiDB Plugin 等手段，使 TiDB 增加 FDW 功能。
+
+## Background
+
+TiDB 访问远程数据源是一个经久不衰的话题。不完全统计，两届 Hackathon 中，有各类远程数据源访问的项目。例如: CSV，Elasticsearch，Prometheus，Presto 等。但这类需求存在：与 TiDB 生态结合度不强，难以进行稳定官方维护，难以合并至 master 分支落地等问题。
+
+本草案结合前人的项目经验，设计了一套 TiDB FDW 方案。同时有极高的落地性和可扩展性，吸引更多人来构建 TiDB 生态。
+
+## Proposal
+
+将 FDW 功能加入到 TiDB 中。严格循序 PostgreSQL FDW 与 [SQL/MED](https://wiki.postgresql.org/wiki/SQL/MED#Version_4)。
+
+用户可以通过以下流程使用 FDW 扩展。
+
+1. 遵守 plugin framework，创建 Engine Plugin。 
+2. 实现 engine plugin 接口（具体参考 Implementation 部分 )
+3. build plugin 
+4. tidb 中加载相应 engine plugin
+
+这里举使用 CSV FDW 作为例子，用户的使用流程为：
+
+```sql
+# 标准 FDW 方式创建 使用 csv engine 的table
+CREATE EXTENSION csv_fdw;
+CREATE SERVER test_logs FOREIGN DATA WRAPPER csv_fdw;
+CREATE FOREIGN TABLE people (city int, name char(255)) SERVER test_logs
+OPTIONS ( filename '<YOUR_CSV_FILE_PATH>');
+
+# 使用 MySQL 兼容方案创建 table
+CREATE TABLE people(city int, name char(255)) ENGINE = csv ENGINE_ATTRIBUTE = "path=<YOUR_CSV_FILE_PATH>"
+  
+# 支持 insert  
+INSERT INTO people values(1, 'wph95');
+INSERT INTO people values(3, 'iguory');
+INSERT INTO people values(0, 'lfkdsk');
+
+# 支持 select
+SELECT * FROM people;
+SELECT * FROM people where city = 2;
+
+# 支持与 TiDB 普通 table: city 进行 join 
+SELECT city.cityName, people.name FROM city INNER JOIN people ON city.id=people.city
```

Comment:
This won't be very special even if we just support table scan. The complex part might be if we support join within FD engine if it is a database.